vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js

```diff
@@ -1,1017 +1,1023 @@
+         * First mapping the item to physical (iron list) indexes, so that we update
+         * only items in with the correct index that are cached in the iron list.
+         */
+        const itemKeys = items.map(item => item.key);
+        const indexes = grid._physicalItems
+            .map((tr, index) => tr._item && tr._item.key && itemKeys.indexOf(tr._item.key) > -1 ? index : null)
+            .filter(idx => idx !== null);
+        if (indexes.length > 0) {
+          grid._assignModels(indexes);
+        }
+      });
 
-      for (let i = 0; i < updatedPageCount; i++) {
-        let page = firstPage + i;
-        let slice = items.slice(i * grid.pageSize, (i + 1) * grid.pageSize);
-        if(!cache[pkey]) {
-          cache[pkey] = {};
+      grid.$connector.set = tryCatchWrapper(function(index, items, parentKey) {
+        if (index % grid.pageSize != 0) {
+          throw 'Got new data to index ' + index + ' which is not aligned with the page size of ' + grid.pageSize;
         }
-        cache[pkey][page] = slice;
+        let pkey = parentKey || root;
+
+        const firstPage = index / grid.pageSize;
+        const updatedPageCount = Math.ceil(items.length / grid.pageSize);
 
-        grid.$connector.doSelection(slice.filter(
-          item => item.selected && !isSelectedOnGrid(item)));
-        grid.$connector.doDeselection(slice.filter(
-          item => !item.selected  && (selectedKeys[item.key] || isSelectedOnGrid(item))));
+        for (let i = 0; i < updatedPageCount; i++) {
+          let page = firstPage + i;
+          let slice = items.slice(i * grid.pageSize, (i + 1) * grid.pageSize);
+          if(!cache[pkey]) {
+            cache[pkey] = {};
+          }
+          cache[pkey][page] = slice;
+
+          grid.$connector.doSelection(slice.filter(
+            item => item.selected && !isSelectedOnGrid(item)));
+          grid.$connector.doDeselection(slice.filter(
+            item => !item.selected  && (selectedKeys[item.key] || isSelectedOnGrid(item))));
 
-        const updatedItems = updateGridCache(page, pkey);
-        if (updatedItems) {
-          itemsUpdated(updatedItems);
-          updateGridItemsInDomBasedOnCache(updatedItems);
+          const updatedItems = updateGridCache(page, pkey);
+          if (updatedItems) {
+            itemsUpdated(updatedItems);
+            updateGridItemsInDomBasedOnCache(updatedItems);
+          }
         }
-      }
-    };
-
-    const itemToCacheLocation = function(item) {
-      let parent = item.parentUniqueKey || root;
-      if(cache[parent]) {
-        for (let page in cache[parent]) {
-          for (let index in cache[parent][page]) {
-            if (grid.getItemId(cache[parent][page][index]) === grid.getItemId(item)) {
-              return {page: page, index: index, parentKey: parent};
+      });
+
+      const itemToCacheLocation = tryCatchWrapper(function(item) {
```

Comment:
This seems to only be called from within other methods that are already using `tryCatchWrapper` so it could be removed here.

Reply:
Updated.